src/sleep/sleep.c

```diff
@@ -290,6 +290,20 @@ static int execute_s2h(const SleepConfig *sleep_config) {
                 else
                         return log_error_errno(r, "Error fetching battery capacity percentage: %m");
 
+                r = get_battery_discharge_rate();
+                if ((last_capacity * 2) <= r)
+                        break;
+                        /* System should hibernate in case discharge rate is higher than double of battery current capacity
+                         * why double : Because while calculating suspend interval, we have taken a buffer of 30 minute and
+                         * discharge_rate is calculated on per 60 minute basis which is double. Also suspend_interval > 0 */
+                else if (r > 0) {
+                        log_debug("Estimating suspend interval using stored discharge rate");
+                        previous_discharge_rate = r;
+                        suspend_interval = ((last_capacity / previous_discharge_rate) * 60 - 30) * USEC_PER_MINUTE;
```

Comment:
`previous_discharge_rate` may be 0, no?? If not, please add an assertion about that.

Also, redundant `()`.

Also, `last_capacity / previous_discharge_rate * 60 - 30` may be negative... Please use `usec_sub_unsigned()`.
```
usec_sub_unsigned(last_capacity / previous_discharge_rate * USEC_PER_HOUR, 30 * USEC_PER_MIN);
```

Reply:
> `previous_discharge_rate` may be 0, no?? If not, please add an assertion about that.
> 
fixed.

> Also, redundant `()`.
> 
fixed.
> Also, `last_capacity / previous_discharge_rate` may easily truncated to 0, and `last_capacity / previous_discharge_rate * 60 - 30` may be negative... Please use `usec_sub_unsigned()`.
> 
> ```
> usec_sub_unsigned(last_capacity * USEC_PER_HOUR / previous_discharge_rate, 30 * USEC_PER_MIN);
> ```

No it will not be negative. In fact, we do not need the suspend_interval if it is going to be negative. The value will not be valid to use so this has to be positive. Also 
`last_capacity * 2 <= r` will make sure that suspend interval is always positive.

Reply:
you are right here. i will fix this.
```
Jul 12 23:18:26 localhost systemd-sleep[560]: Battery is not low
Jul 12 23:18:26 localhost systemd-sleep[560]: Current battery charge percentage: 13%
Jul 12 23:18:26 localhost systemd-sleep[560]: Stored hash_id and discharge_rate is c89f590e649c5053f18fd1c4a56ca0b8e1270cb4e501386406adfbca902e59fb 25
Jul 12 23:18:26 localhost systemd-sleep[560]: [/sys](tg://bot_command?command=sys)[/class](tg://bot_command?command=class)[/power_supply](tg://bot_command?command=power_supply)[/BAT0](tg://bot_command?command=BAT0)[/manufacturer](tg://bot_command?command=manufacturer) is unavailable: No such file or directory
Jul 12 23:18:26 localhost systemd-sleep[560]: [/sys](tg://bot_command?command=sys)[/class](tg://bot_command?command=class)[/power_supply](tg://bot_command?command=power_supply)[/BAT0](tg://bot_command?command=BAT0)[/model_name](tg://bot_command?command=model_name) is unavailable: No such file or directory
Jul 12 23:18:26 localhost systemd-sleep[560]: [/sys](tg://bot_command?command=sys)[/class](tg://bot_command?command=class)[/power_supply](tg://bot_command?command=power_supply)[/BAT0](tg://bot_command?command=BAT0)[/serial_number](tg://bot_command?command=serial_number) is unavailable: No such file or directory
Jul 12 23:18:26 localhost systemd-sleep[560]: product_id does not exist: No such file or directory
Jul 12 23:18:26 localhost systemd-sleep[560]: stored hash_id is c89f590e649c5053f18fd1c4a56ca0b8e1270cb4e501386406adfbca902e59fb and stored discharge_rat
e is 25
Jul 12 23:18:26 localhost systemd-sleep[560]: hash id match
Jul 12 23:18:26 localhost systemd-sleep[560]: Battery discharge rate is in valid range
Jul 12 23:18:26 localhost systemd-sleep[560]: Estimating suspend interval using stored discharge rate
Jul 12 23:18:26 localhost systemd-sleep[560]: Set timerfd wake alarm for 584542y 2w 2d 19h 31min 49s
```

Reply:
fixed